File: src/test/java/org/vanautrui/languages/compiler/parsing/astnodes/nonterminals/upperscopes/StructDeclNodeTest.java
package org.vanautrui.languages.compiler.parsing.astnodes.nonterminals.upperscopes;

import org.junit.Assert;
import org.junit.Test;
import org.vanautrui.languages.compiler.lexing.Lexer;
import org.vanautrui.languages.compiler.lexing.utils.TokenList;
import org.vanautrui.languages.compiler.parsing.astnodes.nonterminal.upperscopes.StructDeclNode;

import static org.junit.Assert.assertEquals;

public class StructDeclNodeTest {

  @Test
  public void test_can_parse_empty_struct_decl() throws Exception {

    TokenList list = (new Lexer()).lexCodeTestMode("struct MyStruct{}");

    StructDeclNode node = new StructDeclNode(list,false);

    assertEquals(0,node.structMembersList.size());
    assertEquals("MyStruct",node.getTypeName());
  }

  @Test
  public void test_will_not_parse_invalid_typename_for_struct() throws Exception {

    TokenList list = (new Lexer()).lexCodeTestMode("struct myStruct{}");

    try {
      StructDeclNode node = new StructDeclNode(list, false);
      Assert.fail();
    }catch (Exception e){
      //
    }
  }

  @Test
  public void test_rejects_struct_with_subroutine_type() throws Exception {

    TokenList list = (new Lexer()).lexCodeTestMode("struct ()~>PInt{PInt a}");

    try {
      StructDeclNode node = new StructDeclNode(list, false);
      Assert.fail();
    }catch (Exception e){
      //
    }
  }

  @Test
  public void test_can_parse_struct_with_1_member() throws Exception {

    TokenList list = (new Lexer()).lexCodeTestMode("struct MyStruct{PInt a}");

    StructDeclNode node = new StructDeclNode(list,false);

    assertEquals(1,node.structMembersList.size());
    assertEquals("a",node.structMembersList.get(0).name);
    assertEquals("PInt",node.structMembersList.get(0).type.getTypeName());
    assertEquals("MyStruct",node.getTypeName());
  }

  @Test
  public void test_can_parse_struct_with_2_members() throws Exception {

    TokenList list = (new Lexer()).lexCodeTestMode("struct MyStruct{PInt a,MyStruct b}");

    StructDeclNode node = new StructDeclNode(list,false);

    assertEquals(2,node.structMembersList.size());

    assertEquals("a",node.structMembersList.get(0).name);
    assertEquals("PInt",node.structMembersList.get(0).type.getTypeName());

    assertEquals("b",node.structMembersList.get(1).name);
    assertEquals("MyStruct",node.structMembersList.get(1).type.getTypeName());

    assertEquals("MyStruct",node.getTypeName());
  }
}

File: dragon-parser/src/test/astnodes/nonterminals/statements/upperscopes/StructDeclNodeTest.cpp
package org.vanautrui.languages.test.astnodes.nonterminals.statements.upperscopes;

import org.vanautrui.languages.commandline.ParserPhases;
import org.vanautrui.languages.compiler.lexing.utils.TokenList;
import org.vanautrui.languages.compiler.parsing.astnodes.nonterminal.upperscopes.StructDeclNode;

public final class StructDeclNodeTest {

	public static int test_can_parse_empty_struct_decl() throws Exception {

		final TokenList list = ParserPhases.makeTokenList("struct MyStruct{}", false);

		StructDeclNode node = new StructDeclNode(list, false);

		boolean assert1 = (0 == node.members.size());

		return (assert1)?1:0;
	}

	public static int test_will_not_parse_invalid_typename_for_struct() throws Exception {

		final TokenList list = ParserPhases.makeTokenList("struct myStruct{}", false);

		try {
			final StructDeclNode node = new StructDeclNode(list, false);
			return 0;
		} catch (Exception e) {
			return 1;
		}
	}

	public static int test_rejects_struct_with_subroutine_type() throws Exception {

		final TokenList list = ParserPhases.makeTokenList("struct ()~>PInt{PInt a}", false);

		try {
			StructDeclNode node = new StructDeclNode(list, false);
			return 0;
		} catch (Exception e) {
			return 1;
		}
	}

	public static int test_can_parse_struct_with_1_member() throws Exception {

		final TokenList list = ParserPhases.makeTokenList("struct MyStruct{PInt a}", false);

		final StructDeclNode node = new StructDeclNode(list, false);

		final boolean assert1 = (1 == node.members.size());
		final boolean assert2 = "a".equals( node.members.get(0).name);

		return (assert1&&assert2)?1:0;
	}

	public static int test_can_parse_struct_with_2_members() throws Exception {

		final TokenList list = ParserPhases.makeTokenList("struct MyStruct{PInt a,MyStruct b}", false);

		final StructDeclNode node = new StructDeclNode(list, false);

		final boolean assert1 = (2 == node.members.size());

		final boolean assert2 = "a".equals( node.members.get(0).name);

		final boolean assert3 = "b".equals( node.members.get(1).name);

		return (assert1&&assert2&&assert3)?1:0;
	}
}
